### PR TITLE
6232281: -XX:-UseLoopSafepoints causes assert(v_false,"Parse::remove_useless_nodes missed this node")

### PR DESCRIPTION
**Pull request by robcasloz**
Check for nodes missed by `remove_useless_nodes()` only if PhaseRemoveUseless has actually been run. This makes it possible to use `-XX:-UseLoopSafepoints` without crashing trivially, although implicit assumptions in other parts of C2 about the existence of loop safepoints might lead to more subtle failures for more complex methods.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-6232281](https://bugs.openjdk.java.net/browse/JDK-6232281): -XX:-UseLoopSafepoints causes assert(v_false,"Parse::remove_useless_nodes missed this node")

### Reviewers
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1311/head:pull/1311`
`$ git checkout pull/1311`

### CODE REVIEW

**Comment by robcasloz** [comment]
/summary
Check for nodes missed by remove_useless_nodes() only if PhaseRemoveUseless has
actually been run. This makes it possible to use -XX:-UseLoopSafepoints without
crashing trivially, although implicit assumptions in other parts of C2 about the
existence of loop safepoints might lead to more subtle failures for more complex
methods.

**Comment by robcasloz** [comment]
Tested on hs-tier{1,2,3}.

**Review by neliasso — APPROVED** [review]
Looks good!

File: test/hotspot/jtreg/compiler/arguments/TestDisableUseLoopSafepoints.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * @test
+ * @bug 6232281
+ * @requires vm.debug == true & vm.compiler2.enabled
+ * @summary Tests that C2 does not crash trivially with a "remove_useless_nodes
+ *          missed this node" message when UseLoopSafepoints is disabled. Note
+ *          that other parts of C2 assume that this option is enabled and might
+ *          fail in more subtle ways otherwise.
+ * @run main/othervm -Xcomp -Xbatch -XX:-TieredCompilation
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/arguments/TestDisableUseLoopSafepoints.java:32]: `-Xcomp` implies `-Xbatch`

File: test/hotspot/jtreg/compiler/arguments/TestDisableUseLoopSafepoints.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+/**
+ * @test
+ * @bug 6232281
+ * @requires vm.debug == true & vm.compiler2.enabled
+ * @summary Tests that C2 does not crash trivially with a "remove_useless_nodes
+ *          missed this node" message when UseLoopSafepoints is disabled. Note
+ *          that other parts of C2 assume that this option is enabled and might
+ *          fail in more subtle ways otherwise.
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/arguments/TestDisableUseLoopSafepoints.java:31]: Please make sure that this test does not trigger any of these other failures when executed in the CI.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

File: test/hotspot/jtreg/compiler/arguments/TestDisableUseLoopSafepoints.java (hunk 1 above)
Comment (robcasloz) [reply test/hotspot/jtreg/compiler/arguments/TestDisableUseLoopSafepoints.java]: Thanks, updated!

File: test/hotspot/jtreg/compiler/arguments/TestDisableUseLoopSafepoints.java (hunk 2 above)
Comment (robcasloz) [reply test/hotspot/jtreg/compiler/arguments/TestDisableUseLoopSafepoints.java]: The last part of the comment referred to the case `-XX:-UseLoopSafepoints` is used on more complex methods than the tested one. I have just removed it, as it was confusing and probably doing more harm than good. The test succeeds for all CI debug configurations and is correctly skipped for all CI release configurations.

**Comment by robcasloz** [comment]
Thanks for reviewing Nils and Tobias! I have updated the test with the feedback from Tobias.

**Review by TobiHartmann — APPROVED** [review]

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by navyxliu** [comment]
LGTM. Thanks for fixing it.

**Review by neliasso — APPROVED** [review]

**Comment by robcasloz** [comment]
Thanks for reviewing Vladimir and Xin!

**Pull request closed without merging**